### PR TITLE
Extend Test Coverage for `top`, `parent` HTML 7.1.1 Nested Browsing Contexts

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR adds some additional test coverage to `window` attributes as outlined in [WHATWG HTML 7.1.1 Nested Browsing Contexts](https://html.spec.whatwg.org/#nested-browsing-contexts).

* I've extended `window.top` tests to account for cases in which `top` should return null.
* I've added coverage for `window.parent`, including `null` cases
* I did some minor cleanup and renaming for clarity

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/209366935)

# Firefox (nightly channel) #
Testing web-platform-tests at revision 70be7b912ff837ad60c6b7099a9121df78b7d2d7
Using browser at version BuildID 20170306110339; SourceStamp 966464a68a2cb3ca1125808e34abb5c1d34e3797
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>5 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5105/html/browsers/windows/nested-browsing-contexts/frameElement.html">/html/browsers/windows/nested-browsing-contexts/frameElement.html</a></summary>


|                                                                       Subtest                                                                        | Results | Messages |
|------------------------------------------------------------------------------------------------------------------------------------------------------|---------|----------|
|                                                                                                                                                      | OK      |          |
| `The window's frameElement attribute must return its container element if it is a nested browsing context`                                           | PASS    |          |
| `The SecurityError must be thrown if the window accesses to frameElement attribute of a Window which does not have the same effective script origin` | PASS    |          |
| `The window's frameElement attribute must return null if the container's document does not have the same effective script origin`                    | PASS    |          |
| `The window's frameElement attribute must return null if it is not a nested browsing context`                                                        | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5105/html/browsers/windows/nested-browsing-contexts/window-parent-null.html">/html/browsers/windows/nested-browsing-contexts/window-parent-null.html</a></summary>


|                                   Subtest                                   | Results | Messages |
|-----------------------------------------------------------------------------|---------|----------|
|                                                                             | OK      |          |
| `\`window.parent\` is null when browsing context container element removed` | PASS    |          |
| `\`window.parent\` null when parent browsing context container removed`     | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5105/html/browsers/windows/nested-browsing-contexts/window-parent.html">/html/browsers/windows/nested-browsing-contexts/window-parent.html</a></summary>


|                          Subtest                          | Results | Messages |
|-----------------------------------------------------------|---------|----------|
|                                                           | OK      |          |
| `\`window.parent\` for top-level browsing context`        | PASS    |          |
| `\`window.parent\` on single nested browsing context`     | PASS    |          |
| `\`window.parent\` for multiple nested browsing contexts` | PASS    |          |
| `\`window.parent\` for new, top-level browsing context`   | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5105/html/browsers/windows/nested-browsing-contexts/window-top-null.html">/html/browsers/windows/nested-browsing-contexts/window-top-null.html</a></summary>


|                                  Subtest                                  | Results | Messages |
|---------------------------------------------------------------------------|---------|----------|
|                                                                           | OK      |          |
| `\`window.top\` is null when browsing context container element removed`  | PASS    |          |
| `\`window.top\`null when any ancestor browsing context container removed` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5105/html/browsers/windows/nested-browsing-contexts/window-top.html">/html/browsers/windows/nested-browsing-contexts/window-top.html</a></summary>


|           Subtest            | Results |     Messages     |
|------------------------------|---------|------------------|
|                              | TIMEOUT |                  |
| `Top level browsing context` | PASS    |                  |
| `One nested iframe`          | PASS    |                  |
| `Two nested iframes`         | TIMEOUT | `Test timed out` |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/209366936)

# Chrome (unstable channel) #
Testing web-platform-tests at revision 70be7b912ff837ad60c6b7099a9121df78b7d2d7
Using browser at version 58.0.3029.6 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>5 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5105/html/browsers/windows/nested-browsing-contexts/frameElement.html">/html/browsers/windows/nested-browsing-contexts/frameElement.html</a></summary>


|                                                                       Subtest                                                                        | Results | Messages |
|------------------------------------------------------------------------------------------------------------------------------------------------------|---------|----------|
|                                                                                                                                                      | OK      |          |
| `The window's frameElement attribute must return its container element if it is a nested browsing context`                                           | PASS    |          |
| `The SecurityError must be thrown if the window accesses to frameElement attribute of a Window which does not have the same effective script origin` | PASS    |          |
| `The window's frameElement attribute must return null if the container's document does not have the same effective script origin`                    | PASS    |          |
| `The window's frameElement attribute must return null if it is not a nested browsing context`                                                        | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5105/html/browsers/windows/nested-browsing-contexts/window-parent-null.html">/html/browsers/windows/nested-browsing-contexts/window-parent-null.html</a></summary>


|                                   Subtest                                   | Results | Messages |
|-----------------------------------------------------------------------------|---------|----------|
|                                                                             | OK      |          |
| `\`window.parent\` is null when browsing context container element removed` | PASS    |          |
| `\`window.parent\` null when parent browsing context container removed`     | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5105/html/browsers/windows/nested-browsing-contexts/window-parent.html">/html/browsers/windows/nested-browsing-contexts/window-parent.html</a></summary>


|                          Subtest                          | Results | Messages |
|-----------------------------------------------------------|---------|----------|
|                                                           | OK      |          |
| `\`window.parent\` for top-level browsing context`        | PASS    |          |
| `\`window.parent\` on single nested browsing context`     | PASS    |          |
| `\`window.parent\` for multiple nested browsing contexts` | PASS    |          |
| `\`window.parent\` for new, top-level browsing context`   | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5105/html/browsers/windows/nested-browsing-contexts/window-top-null.html">/html/browsers/windows/nested-browsing-contexts/window-top-null.html</a></summary>


|                                  Subtest                                  | Results | Messages |
|---------------------------------------------------------------------------|---------|----------|
|                                                                           | OK      |          |
| `\`window.top\` is null when browsing context container element removed`  | PASS    |          |
| `\`window.top\`null when any ancestor browsing context container removed` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5105/html/browsers/windows/nested-browsing-contexts/window-top.html">/html/browsers/windows/nested-browsing-contexts/window-top.html</a></summary>


|           Subtest            | Results | Messages |
|------------------------------|---------|----------|
|                              | OK      |          |
| `Top level browsing context` | PASS    |          |
| `One nested iframe`          | PASS    |          |
| `Two nested iframes`         | PASS    |          |

</details>
</details>

**Comment by zcorpan** [comment]
Lint error:
> html/browsers/windows/nested-browsing-contexts/window-top-null.html:44: Whitespace at EOL (TRAILING WHITESPACE)

(Tip: check if your editor has a setting to strip trailing whitespace.)

**Comment by w3c-bots** [comment]
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5105/)

**Comment by lyzadanger** [comment]
(Linting fixed. My apologies. I have an excuse/story for it but doesn't everyone?) :)

**Comment by jgraham** [comment]
@annevk do you have time to review this, or would you like me to take it?

File: html/browsers/windows/nested-browsing-contexts/window-parent-null.html [hunk 1]
```diff
@@ -0,0 +1,71 @@
+<!doctype html>
+<meta charset="utf-8">
+<title>window.parent: `null`</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<div id="log"></div>
+<script>
+
+var t1 = async_test('`window.parent` is null when browsing context container element removed');
+t1.step(function () {
```
Comment (annevk) [inline html/browsers/windows/nested-browsing-contexts/window-parent-null.html]: You can use 
```js
async_test(function (t) {
 ...
 ... = t.step_func(...)
 ...
}, 'description here')
```
instead which is generally clearer then incrementing variable names.

Comment (annevk) [inline html/browsers/windows/nested-browsing-contexts/window-parent-null.html]: You should also feel free to use `async_test(t => { ... }, 'desc')` by the way. Arrow functions are in all browsers.

File: html/browsers/windows/nested-browsing-contexts/window-parent-null.html [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  var iframe = document.createElement('iframe');
+  iframe.onload  = t1.step_func(function () {
+    var iWindow = iframe.contentWindow;
+    assert_equals(iWindow.parent, window);
+    document.body.removeChild(iframe);
+    assert_equals(iWindow.parent, null);
+    t1.done();
```
Comment (annevk) [inline html/browsers/windows/nested-browsing-contexts/window-parent-null.html]: You can use `step_func_done` instead and remove this call.

**Review by annevk — COMMENTED** [review]
Thanks! This generally seems fine to me. I have some remarks:

1. Did you make sure test.html and testcase3.html are not used elsewhere?
2. I noticed you found one timeout case in Firefox, it would be nice to file a bug against that browser for that: https://bugzilla.mozilla.org/enter_bug.cgi?product=Core&component=DOM. This a) gives us an extra check and b) helps them become more interoperable.
3. Other than that I only have some style improvements you might want to use or start using in the future. Not a requirement to make changes for this pull request I think.

File: html/browsers/windows/nested-browsing-contexts/window-parent-null.html (hunk 1 above)
Comment (lyzadanger) [inline html/browsers/windows/nested-browsing-contexts/window-parent-null.html]: For posterity: I agree! It also will keep the non-memorably-named test `var`s out of global namespace. Will update.

**Comment by lyzadanger** [comment]
Good morrow, @annevk ! Turns out there were some interesting things going on beyond stylistic updates (though those are done, too) with the results, and I wanted to take a deeper look at those.

I wanted to understand the failures that were introduced by my new tests (i.e. I didn't take a deep look at pre-existing failures beyond validating that my small changes to those test files didn't _change_ results).

Regarding the Firefox timeouts, this appears to be a manifestation of [this open bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1229707) in which the `load` event on nested `iframe` elements does not fire if the `iframe` does not have a `src`. That's a real bug, but, after deliberation, I decided to add a `src` to the affected `iframe`s (`/common/blank.html`) in my tests and commented them as workarounds. At the time it seemed like a better situation than having most of the tests fail here because of a tangential issue. I am not sure if that's the way you'd like to see this handled, so let me know: I'm more than happy to back that out and leave the tests failing for Firefox. After that change, all of the new tests are passing in Firefox.

However, there are some interesting failures in Safari (and also latest webkit nightly). It would seem that Webkit is holding onto references of `window.parent` even after one of the nested browsing containers (i.e. either the parent or the child) involved is removed from the document. From my understanding of the spec, this should not happen, as the parent relationship with browsing contexts requires both parties involved to be connected and active—`window.parent` should return `null` here. Webkit does not display this behavior with `window.top`; those tests have results I would expect in Webkit. If my interpretation of the [spec language](https://html.spec.whatwg.org/#nested-browsing-contexts) seems off to you, let me know, but otherwise I think I'll open a bug on Webkit this morning.

Finally, note that the history of this branch has become a bit chattery with the updates. When we're getting closer to merge, I can rebase to something more manageable if you'd like!

**Comment by annevk** [comment]
1. Style changes look great! More readable too 😊
1. Since Firefox already has a bug, I don't think we need tests here that fail it necessarily for that reason. Those will be added as part of fixing that bug.
1. Seems good to file a bug on WebKit.
1. If you have a proposed commit message I can use the "Squash and merge" UI from GitHub to clean up the chatter. That's what we usually do and requires a title and a body.

**Comment by lyzadanger** [comment]
> Since Firefox already has a bug, I don't think we need tests here that fail it necessarily for that reason. Those will be added as part of fixing that bug.

Just want to be sure I'm parsing this correctly: I should leave the workaround (that allows these tests to _pass_) in place?

> If you have a proposed commit message I can use the "Squash and merge" UI from GitHub to clean up the chatter. That's what we usually do and requires a title and a body.

This OK?:

```
Expand test coverage for nested browsing contexts

Improve test coverage for WHATWG HTML Section 7.1.1 Nested Browsing Contexts.
Extend tests for `window.top` and add new tests for `window.parent`.
```

**Comment by annevk** [comment]
Yeah, keeping the workaround seems fine so the other aspects get tested (and the workaround will get tested once Firefox fixes the bug by a new test).

Commit message was good too. (I did add a HTML prefix to the title for extra clarity.)

Thanks!

**Comment by annevk** [comment]
Oh and if you could leave a comment here with a pointer to the WebKit bug once you have filed it that might be nice in case someone stumbles upon this thread later.

**Comment by lyzadanger** [comment]
See https://bugs.webkit.org/show_bug.cgi?id=169467

**Pull request merged**